// @trayio/commons/dist/dynamictype/DynamicType.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.dynamicArrayTypeDescriptor = exports.dynamicObjectTypeDescriptor = exports.dynamicTypeDescriptor = exports.DynamicType = exports.DynamicObject = void 0;
const TE = __importStar(require("fp-ts/TaskEither"));
const fs = __importStar(require("fs"));
const fse = __importStar(require("fs-extra"));
const E = __importStar(require("fp-ts/Either"));
const function_1 = require("fp-ts/function");
const t = __importStar(require("io-ts"));
const Base64_1 = require("../base64/Base64");
const JsonSerialization_1 = require("../serialization/JsonSerialization");
const Try_1 = require("../try/Try");
const RemoveNullValuesCodec_1 = require("../codec/RemoveNullValuesCodec");
const serializer = new JsonSerialization_1.JsonSerialization();
const testRemoveNullValuesCodec = RemoveNullValuesCodec_1.RemoveNullValuesCodec.instance;
exports.DynamicObject = {
    fromBase64String: (base64Str) => (0, function_1.pipe)(Base64_1.Base64.fromBase64String(base64Str), E.map((base64) => Base64_1.Base64.decodeBinary(base64)), E.chain((decoded) => serializer.deserialize(decoded)), E.map((decoded) => decoded)),
    removeNullValues: (value) => testRemoveNullValuesCodec.encode(value),
};
exports.DynamicType = {
    readFromFile: (path) => (0, Try_1.tryToEither)(() => JSON.parse(fs.readFileSync(path, 'utf8'))),
    writeToFile: (path, value) => (0, function_1.pipe)((0, Try_1.tryToEither)(() => fs.writeFileSync(path, JSON.stringify(value))), E.map(() => undefined)),
    writeToFileAsync: (path, value) => (0, function_1.pipe)(TE.tryCatch(() => fse.writeFile(path, JSON.stringify(value)), (err) => new Error(err.message)), TE.map(() => undefined)),
    safeCast: (dynamicType, codec) => codec.decode(dynamicType),
};
exports.dynamicTypeDescriptor = t.recursion('DynamicType', () => t.union([
    t.string,
    t.number,
    t.boolean,
    t.null,
    exports.dynamicObjectTypeDescriptor,
    exports.dynamicArrayTypeDescriptor,
]));
exports.dynamicObjectTypeDescriptor = t.recursion('DynamicObject', () => t.record(t.string, exports.dynamicTypeDescriptor));
exports.dynamicArrayTypeDescriptor = t.recursion('DynamicArray', () => t.array(exports.dynamicTypeDescriptor));
