// @trayio/commons/dist/codec/RemoveNullValuesCodec.unit.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const RemoveNullValuesCodec_1 = require("./RemoveNullValuesCodec");
const testRemoveNullValuesCodec = RemoveNullValuesCodec_1.RemoveNullValuesCodec.instance;
describe('RemoveNullValuesCodec', () => {
    test('should remove null values', () => {
        const value = {
            firtName: 'blah',
            lastName: null,
            addresses: [
                {
                    street: 'blah street',
                    number: 123,
                    city: {
                        name: 'London',
                        province: null,
                        state: null,
                    },
                    notes: 'blah blah blah',
                },
                {
                    street: 'blah blah street',
                    number: 321,
                    city: {
                        name: 'New York',
                        province: null,
                        state: 'NY',
                    },
                    notes: null,
                },
            ],
            phone: {
                areaCode: null,
                number: 1234567,
            },
        };
        const actualValue = testRemoveNullValuesCodec.encode(value);
        const expectedValue = {
            firtName: 'blah',
            addresses: [
                {
                    street: 'blah street',
                    number: 123,
                    city: {
                        name: 'London',
                    },
                    notes: 'blah blah blah',
                },
                {
                    street: 'blah blah street',
                    number: 321,
                    city: {
                        name: 'New York',
                        state: 'NY',
                    },
                },
            ],
            phone: {
                number: 1234567,
            },
        };
        expect(actualValue).toEqual(expectedValue);
    });
});
